File: src/com/medium/FindBottomLeftTreeValue_513.java
/**
 * 
 */
package com.medium;

import java.util.LinkedList;
import java.util.Queue;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class FindBottomLeftTreeValue_513 {
    // Use traversal DRL
    // Time Complexity: O(n)
    // Space Complexity: O(n)
    // Runtime: 1 ms, faster than 69.84%
    // Memory Usage: 39 MB, less than 86.20%
    private TreeNode mNode;

    public int findBottomLeftValue(TreeNode root) {
        if (root == null) {
            return 0;
        }
        Queue<TreeNode> mQueue = new LinkedList<>();
        mQueue.add(root);

        // D -> R -> L, then last is the leftmost node
        while (mQueue.isEmpty() == false) {
            mNode = mQueue.poll();

            if (mNode.right != null) {
                mQueue.add(mNode.right);
            }
            if (mNode.left != null) {
                mQueue.add(mNode.left);
            }
        }
        return mNode.val;

    }
}

File: Golang_src/medium/findBottomLeftTreeValue_513.go
package medium

// TreeNode for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// Use BFS:
// Time Complexity: O(n)
// Space Complexity:O(h)
// Runtime: 4 ms, faster than 100.00%
// Memory Usage: 5.7 MB, less than 70.59%
func findBottomLeftValue(root *TreeNode) int {
	var node *TreeNode
	queue := []*TreeNode{root}

	for len(queue) > 0 {
		size := len(queue)

		for i := 0; i < size; i++ {
			node = queue[0]
			queue = queue[1:]

			if node.Right != nil {
				queue = append(queue, node.Right)
			}
			if node.Left != nil {
				queue = append(queue, node.Left)
			}
		}
	}

	return node.Val
}
